Add description and initial guidance for DOTNET_HOST_PATH usage

baronfel opened the pull request:
## Summary

Based on conversations in https://github.com/dotnet/runtime/issues/88754, https://github.com/dotnet/roslyn/pull/69186, and other places, we need documentation about this environment variable and its purpose. This is a draft of that documentation - though we may also need a dedicated page for "How should I locate the correct `dotnet` binary?" for application developers.


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/tools/dotnet-environment-variables.md](https://github.com/dotnet/docs/blob/c3470c9f8a3ef23169dbf71c13bd5c13ad98bb5f/docs/core/tools/dotnet-environment-variables.md) | [.NET environment variables](https://review.learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables?branch=pr-en-us-36383) |


<!-- PREVIEW-TABLE-END -->

baronfel committed:
add description and initial guidance for DOTNET_HOST_PATH usage

baronfel committed:
update note for apphost-based-execution

eerhardt commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+As a tool author, if you want to locate the correct `dotnet` binary to use to spawn a .NET binary, you should use the following algorithm:
+
+* if `DOTNET_HOST_PATH` is set, use that value directly
+* otherwise, probe `PATH` according to the rules of your operating system to find the first `dotnet` binary on the `PATH`
```
I wouldn't recommend probing `PATH` according to the rules of your OS. Instead I would suggest to just launch `dotnet` (or `dotnet.exe` depending on OS) and let the OS take care of locating the executable.

cc @vitek-karas

baronfel replied:
That only works if you're using `UseShellExecute=true` though, if I'm remembering right - and many folks spawn commands and redirect the stdin/stdout/stderr which blocks that execution pathway.

vitek-karas replied:
In that case we should pull in security folks - I remember that parsing `PATH` is considered a bad idea, especially on Windows.

Otherwise I completely agree with Eric that we should let the OS sort it out.

jaredpar replied:
At an MSBuild task level you can also just set the tool exe to `dotnet` and let MSBuild probe `$PATH` for you (that is how it works today).

richlander replied:
This document should not be discussing `PATH` at all. It should describe how this ENV is intended to be used, when it is typically set by others, and if/when you (the user) should set it.

eerhardt replied:
> I think we need to discuss what to do when this value is not set.

My recommendation/thought/opinion is above. Copied here:

> Instead I would suggest to just launch dotnet (or dotnet.exe depending on OS) and let the OS take care of locating the executable.

jaredpar replied:
Okay but how does one do that in the context of an MSBuild task? What specifically do we want authors to do here? Cause in the case of `ToolTask` there isn't an option of "let OS take care of it". 

I know that as an implementation detail, you can achieve this by providing the tool path to be simply `dotnet` and then MSBuild will do the work for you. That is not documented AFAIK though.

eerhardt replied:
I believe https://github.com/dotnet/msbuild/issues/9018 is an attempt to try to remedy this for MSBuild. Honestly, I have no idea why MSBuild parses `$PATH` itself, I'm sure it was to work around some corner-case issue. It feels unnecessary. cc @rainersigwald. 

> Okay but how does one do that in the context of an MSBuild task? What specifically do we want authors to do here?

Just say `dotnet` or `dotnet.exe`, as necessary.

jaredpar replied:
> Just say dotnet or dotnet.exe, as necessary.

Then we need to put this in guidance, not ignore it.

rainersigwald replied:
> Just say `dotnet` or `dotnet.exe`, as necessary.

Is that . . . ok? Wouldn't it run different bits in some circumstances, like using a specific old SDK while a newer SDK/runtime is installed?

eerhardt replied:
> Wouldn't it run different bits in some circumstances, like using a specific old SDK while a newer SDK/runtime is installed?

Can you describe the situation you are thinking about more? I'm not following what you'd think would be a problem here, over parsing `$PATH` yourself.

Note if you are already inside a .NET SDK command (like `dotnet build`), then `DOTNET_HOST_PATH` will already be set, so you won't fall into this case where `DOTNET_HOST_PATH` isn't set.

rainersigwald replied:
You're right, I think that completely avoids my concern. In that case we can simplify the MSBuild helper to "value of `DOTNET_HOST_PATH` or `dotnet`/`dotnet.exe`".

baronfel replied:
MSBuild has an issue to investigate this now, and the post has been updated.

eerhardt commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+This environment variable is set by default when the .NET SDK is invoked (typically by using `dotnet build`, `dotnet publish`, and so on). If your application needs to communicate the location of the `dotnet` binary that ran it in a similar way to the SDK, you should also set this environment variable.
+
+> [!NOTE]
+> This environment variable will not be set for apphost-based execution (applications that are directly launched via a binary like `myapp.exe` or `myapp`) instead of through the `dotnet` binary (like `dotnet myapp.dll`). In such cases, the spawned applications should locate `dotnet` via the `PATH` as described above.
```
This line makes it sound like invoking `dotnet myapp.dll` will set this environment variable for any child process. But it won't.  The only place that sets this variable is inside the .NET SDK.

baronfel committed:
make it clear that invoking via dotnet doesn't set the DOTNET_HOST_PATH variable

marcpopMSFT commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+
+Without this environment variable, if a `dotnet` binary was used from a location not on the `PATH`, child processes of that `dotnet` that wished to invoke `dotnet` would invoke a different `dotnet` binary than the one that was used to spawn them - this can result in confusing behavior for end users.
+
+As a tool author, if you want to locate the correct `dotnet` binary to use to spawn a .NET binary, you should use the following algorithm:
```
Is this only used for child-processes of the launched dotnet? In other words, the initial launch of dotnet is still controlled by the path (when typing dotnet) or the customer (when specifying the full path to dotnet)

baronfel replied:
it's a weird scenario yeah - 

* user invokes `dotnet` binary via absolute path or implicitly by their shell discovering the `dotnet` binary on PATH
* `dotnet` loads SDK
* SDK sets path to the user-selected `dotnet` to `DOTNET_HOST_PATH`
* tools in the SDK that need to execute a standalone `dotnet` command use `DOTNET_HOST_PATH`.

so the value only directly impacts those subprocesses, which I will make clearer.

vitek-karas commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+* if `DOTNET_HOST_PATH` is set, use that value directly
+* otherwise, probe `PATH` according to the rules of your operating system to find the first `dotnet` binary on the `PATH`
+
+This environment variable is set by default when the .NET SDK is invoked (typically by using `dotnet build`, `dotnet publish`, and so on). If your application needs to communicate the location of the `dotnet` binary that ran it in a similar way to the SDK, you should also set this environment variable.
```
My understanding is that for SDK we run all the tools with the architecture of the SDK, even if they're targeting different architecture. So this should not be a big problem.

If we needed to solve this for multi-arch correctly we would probably need to introduce things like `DOTNET_HOST_PATH_<arch>`, just like we have `DOTNET_ROOT_<arch>`.

baronfel replied:
Since this is just documenting an existing environment variable and its intended use I'm going to go with documenting the constraint that the `dotnet` in `DOTNET_HOST_PATH` has been 'fixed' to a specific architecture and provides no information for looking up another architecture's `dotnet` binary.

vitek-karas commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+This environment variable is set by default when the .NET SDK is invoked (typically by using `dotnet build`, `dotnet publish`, and so on). If your application needs to communicate the location of the `dotnet` binary that ran it in a similar way to the SDK, you should also set this environment variable.
+
+> [!NOTE]
+> This environment variable will *not* be set for apphost-based execution (applications that are directly launched via a binary like `myapp.exe` or `myapp`) or through applications that are run via the `dotnet` binary (like `dotnet myapp.dll`). In such cases, the spawned applications should locate `dotnet` via the `PATH` as described above.
```
Eventually we should also describe the recommended setup for running "apps" as child processes. Outside of SDK we're pushing people to use executables as the app's entry point, and to avoid the `dotnet app.dll` gesture. (It is problematic for UI apps, and just doesn't work at all for single-file apps). So if we envision usage of executable entrypoitns in this chain then `DOTNET_HOST_PATH` is not going to solve anything.

vitek-karas commented:
+ @MarcoRossignoli who was solving a very similar problem for `dotnet test`.

MarcoRossignoli commented:
>  though we may also need a dedicated page for "How should I locate the correct dotnet binary?"

@baronfel in the test repo we have the alg to find out correct dotnet muxer https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs#L121 
maybe we could think to ship something open like the msbuildlocator for devs.

richlander commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+
+Specifies the absolute path to a `dotnet` binary (`dotnet.exe` on Windows, `dotnet` on non-Windows) that was used to launch this session. This is typically used by a hosting application that wants to ensure that the exact same version of the host binary is used for all child processes.
+
+Without this environment variable, if a `dotnet` binary was used from a location not on the `PATH`, child processes of that `dotnet` that wished to invoke `dotnet` would invoke a different `dotnet` binary than the one that was used to spawn them - this can result in confusing behavior for end users.
```
I don't think paragraph is adding much. It also seems in conflict with the paragraph just before it.

baronfel replied:
Removed it for clarity

baronfel committed:
rework based on PR and internal feedback

clarify that this impacts subprocesses only, and is very focused on the .NET SDK, not a general mechanism

eerhardt commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,18 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+Tools that invoke `dotnet` during an SDK command should use the following algorithm to locate it:
+
+* if `DOTNET_HOST_PATH` is set, use that value directly
+* otherwise, locate `dotnet` on the system's `PATH`
```
```suggestion
* otherwise, use the `dotnet` on the system's `PATH`
```

I think that simple change would address my concerns.

baronfel committed:
Update docs/core/tools/dotnet-environment-variables.md

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

eerhardt reviewed (approved):
LGTM. Thanks @baronfel!

richlander commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,18 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
 
+### `DOTNET_HOST_PATH`
+
+Specifies the absolute path to a `dotnet` host (`dotnet.exe` on Windows, `dotnet` on non-Windows) that was used to launch the currently-running `dotnet` process. This is used by the .NET SDK to help tools that run during .NET SDK commands ensure they use the same `dotnet` runtime for any child `dotnet` processes they create for the duration of the command. Tools and MSBuild Tasks within the SDK t ...
```
"non-Windows" - Can you use another term than that? It gives me the feels of an older Microsoft. How about (and I'm going on a limb here) "on Linux and macOS".

baronfel replied:
sure - I was going for maximally-inclusive terms here but I can see your point.

baronfel replied:
```suggestion
Specifies the absolute path to a `dotnet` host (`dotnet.exe` on Windows, `dotnet` on Linux and macOS) that was used to launch the currently-running `dotnet` process. This is used by the .NET SDK to help tools that run during .NET SDK commands ensure they use the same `dotnet` runtime for any child `dotnet` processes they create for the duration of the command. Tools and MSBuild Tasks within the SDK that invoke binaries via the `dotnet` host are expected to honor this environment variable to ensure a consistent experience.
```

richlander replied:
```suggestion
Specifies the absolute path to a `dotnet` host (`dotnet.exe` on Windows, `dotnet` on Linux and macOS) that was used to launch the currently-running `dotnet` process. This is used by the .NET SDK to  ensure that tools (and any processes they create) that run as part of .NET SDK commands use the same `dotnet`. Tools and MSBuild Tasks within the SDK that invoke .NET apps via the `dotnet` host are expected to honor this environment variable to ensure a consistent experience.
```

baronfel committed:
Update docs/core/tools/dotnet-environment-variables.md

richlander commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,18 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+Tools that invoke `dotnet` during an SDK command should use the following algorithm to locate it:
+
+* if `DOTNET_HOST_PATH` is set, use that value directly
+* otherwise, use the `dotnet` on the system's `PATH`
```
```suggestion
* otherwise, rely on `dotnet` via the system's `PATH`
```

richlander commented on docs/core/tools/dotnet-environment-variables.md:
```diff
@@ -183,6 +183,18 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
+* otherwise, use the `dotnet` on the system's `PATH`
+
+> [!NOTE]
+> `DOTNET_HOST_PATH` is not a general solution for locating the `dotnet` host. It is only intended to be used by tools that are invoked by the .NET SDK during .NET SDK commands.
```
```suggestion
> `DOTNET_HOST_PATH` is not a general solution for locating the `dotnet` host. It is only intended to be used by tools that are invoked by the .NET SDK.
```

baronfel committed:
Update docs/core/tools/dotnet-environment-variables.md

Co-authored-by: Rich Lander <rlander@microsoft.com>

baronfel committed:
Update docs/core/tools/dotnet-environment-variables.md

Co-authored-by: Rich Lander <rlander@microsoft.com>
